src/main.rs

```diff
@@ -78,127 +49,120 @@ fn main() {
+                                }
+                            }
+                        }
 
-    poll.register(&EventedFd(&fd), STATE_CHANGE, Ready::readable(), PollOpt::empty())
-        .expect("Failed to register state_change fd");
-    poll.register(&EventedFd(&timer.as_raw_fd()), TIMER, Ready::readable(), PollOpt::empty())
-        .expect("Failed to register timer fd");
+                        // at this point the fd is ready
+                        match guard.read() {
+                            Ok(_) => dispatch(),
+                            // if we are still "wouldblock", just continue and retry.
+                            Err(WaylandError::Io(e)) if e.kind() == ErrorKind::WouldBlock => {
+                                continue
+                            }
+                            Err(e) => {
+                                panic!("read wayland socket err: {:?}", e);
+                            }
+                        };
+                    }
+                    None => dispatch(),
+                }
+            }
+        });
+    }
 
     println!("### Taking client locks");
     let host = "localhost";
     let port = match testing {
         true => "5666",
-        false => "5600"
+        false => "5600",
     };
-    let _window_lock = singleinstance::get_client_lock(&format!("aw-watcher-window-at-{}-on-{}", host, port)).unwrap();
-    let _afk_lock = singleinstance::get_client_lock(&format!("aw-watcher-afk-at-{}-on-{}", host, port)).unwrap();
+    let _window_lock =
+        singleinstance::get_client_lock(&format!("aw-watcher-window-at-{}-on-{}", host, port))
+            .unwrap();
+    let _afk_lock =
+        singleinstance::get_client_lock(&format!("aw-watcher-afk-at-{}-on-{}", host, port))
+            .unwrap();
 
     println!("### Creating aw-client");
-    let client = aw_client_rust::AwClient::new(host, port, "aw-watcher-wayland");
+    let client = aw_client_rust::blocking::AwClient::new(host, port, "aw-watcher-wayland");
     let hostname = gethostname::gethostname().into_string().unwrap();
     let window_bucket = format!("aw-watcher-window_{}", hostname);
     let afk_bucket = format!("aw-watcher-afk_{}", hostname);
-    client.create_bucket_simple(&window_bucket, "currentwindow")
+    client
+        .create_bucket_simple(&window_bucket, "currentwindow")
         .expect("Failed to create window bucket");
-    client.create_bucket_simple(&afk_bucket, "afkstatus")
+    client
+        .create_bucket_simple(&afk_bucket, "afkstatus")
         .expect("Failed to create afk bucket");
 
     println!("### Watcher is now running");
-    let mut events = Events::with_capacity(1);
-    let mut prev_window : Option<current_window::Window> = None;
     loop {
-        poll.poll(&mut events, None).expect("Failed to poll fds");
-        for event in &events {
-            match event.token() {
-                STATE_CHANGE => {
```

Comment:
I wonder why do we need to send heartbeat if some state changes? I personally want the watcher to ignore the fluctuations when I switch windows very quickly..

Reply:
No need, it just makes the tracking more accurate. Essentially it makes the tracking almost perfect instead of approximate.
Considering how cheap it is to make a heartbeat, I found it to be a nice feature for barely any cost.